File: 03-RECURSION/PHP/reverse-linked-list.php
<?php

# https://leetcode.com/problems/reverse-linked-list/
class Solution
{
    public function reverseList(?ListNode $head): ?ListNode
    {
        if (!$head or !$head->next) {
            return $head;
        }

        $new_head = $this->reverseList($head->next);

        $head->next->next = $head;
        $head->next = null;

        return $new_head;
    }
}

File: 02-LINKED-LISTS/Go/reverse-linked-list.go
package main

type ListNode struct {
	Next *ListNode
	Prev *ListNode
}

// https://leetcode.com/problems/design-browser-history/
func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	curr := head
	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}

	return prev
}
